### PR TITLE
Proposed patches to stable 0.7 branch

### PR DESCRIPTION
**Pull request by aerusso**
The PR needed section on 0.7.3 projects page listed some patches. I've attempted to resolve conflicts and get them able to merge onto the stable branch.

I'm half doing this for myself. I figured putting it here might avoid any duplicated effort, and centralize a discussion for the patches to the stable branch. If this is a bad idea, or a waste of effort, let me know.

### Description

I've rebased the patches in 0.7.3 project onto the current stable head. In particular zfsonlinux/zfs@aee1dd4d983c64db3c3155290d48f05243e85709 touches code that has changed somewhat. @loli10K could you please make sure I've not made a terrible mistake here?

### Motivation and Context

Some of the bugs addressed in these patches sound really annoying, and I'd like to avoid them if possible.

### How Has This Been Tested?

I've built this, and installed it on a machine. I would like to use the build infrastructure here for more tests. I'll be putting it on more machines gradually.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

**Comment by behlendorf** [comment]
@aerusso thanks.  You'll want to open a matching SPL PR with 0cefc9dbcd9de8a35c51e172edabf2f2ecf15f92 cherry-picked to resolve the build issue.  Then add `Requires-spl: refs/pull/123/head` to the top most commit as described [here](https://github.com/zfsonlinux/zfs/wiki/Buildbot-Options#build-spl-from-a-specific-pull-request)

File: tests/zfs-tests/tests/functional/rsend/rsend_024_pos.ksh [hunk 1]
```diff
@@ -49,8 +49,8 @@ streamfs=$POOL/stream
 
 log_onexit resume_cleanup $sendfs $streamfs
 
-test_fs_setup $sendfs $recvfs
-log_must zfs unmount $sendfs
+test_fs_setup $sendfs $recvfs $streamfs
+log_must zfs unmount -f $sendfs
```
Comment (loli10K) [inline tests/zfs-tests/tests/functional/rsend/rsend_024_pos.ksh]: This is from e3bdcb8ad88e7a90726193c9afe96a713729c928, not aee1dd4d983c64db3c3155290d48f05243e85709: maybe we should also add that commit to the stack?

Comment (aerusso) [reply tests/zfs-tests/tests/functional/rsend/rsend_024_pos.ksh]: Sharp eye. I'll remove that.

**Comment by aerusso** [comment]
@Fabian-Gruenbichler I forgot to mention that your commit zfsonlinux/zfs@48fbb9ddbf2281911560dfbc2821aa8b74127315 touched `receive_object_range` in `dmu_send.c`, which, from my very cursory inspection of its history, was added to support encryption, and won't be included in this stable branch.

I only superficially understand zfsonlinux/zfs@48fbb9ddbf2281911560dfbc2821aa8b74127315, so I'm not 100% sure it makes sense to just ignore the change made to `receive_object_range`. Could you please double check me here?

**Comment by Fabian-Gruenbichler** [comment]
On Thu, Oct 12, 2017 at 02:59:12AM -0700, Antonio Russo wrote:
> @Fabian-Gruenbichler I forgot to mention that your commit zfsonlinux/zfs@48fbb9ddbf2281911560dfbc2821aa8b74127315 touched `receive_object_range` in `dmu_send.c`, which, from my very cursory inspection of its history, was added to support encryption, and won't be included in this stable branch.
> 
> I only superficially understand zfsonlinux/zfs@48fbb9ddbf2281911560dfbc2821aa8b74127315, so I'm not 100% sure it makes sense to just ignore the change made to `receive_object_range`. Could you please double check me here?

yes, dropping that hunk is the correct way to go for backporting to 0.7.

**Comment by aerusso** [comment]
Per the updates to [PR 0.7.3](https://github.com/zfsonlinux/zfs/projects/12), I've included e0922b042 and 085b501fb. This should also be pulling the correct SPL now, per your comment, @behlendorf.

**Comment by tonyhutter** [comment]
@aerusso thanks for doing the legwork on this.  I'll look into pulling it in once buildbot finishes.

**Comment by tonyhutter** [comment]
@aerusso everything here looks fine.  I just merged in your 0.7.3 SPL commit, so you can remove  "Minimal requirement for SPL build compatibility".  After that I can merge in this PR.

**Comment by tonyhutter** [comment]
I ended up just cherry picking these and pushed them to `zfs-0.7-release`.  Thanks again @aerusso for putting the patch set together.

**Pull request closed without merging**